spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/GqlDatastoreQuery.java

```diff
@@ -167,43 +206,156 @@ boolean isNonEntityReturnedType(Class returnedType) {
-		for (int i = 0; i < parameters.getNumberOfParameters(); i++) {
-			Parameter param = parameters.getParameter(i);
-			Optional<String> paramName = param.getName();
-			if (!paramName.isPresent()) {
-				throw new DatastoreDataException(
-						"Query method has a parameter without a valid name: "
-								+ getQueryMethod().getName());
+	private List<String> getOriginalParamTags() {
+		if (this.originalParamTags == null) {
+			this.originalParamTags = new ArrayList<>();
+			Set<String> seen = new HashSet<>();
+			Parameters parameters = getQueryMethod().getParameters();
+			for (int i = 0; i < parameters.getNumberOfParameters(); i++) {
+				Parameter param = parameters.getParameter(i);
+				Optional<String> paramName = param.getName();
+				if (!paramName.isPresent()) {
+					throw new DatastoreDataException(
+							"Query method has a parameter without a valid name: "
+									+ getQueryMethod().getName());
+				}
+				String name = paramName.get();
+				if (seen.contains(name)) {
+					throw new DatastoreDataException(
+							"More than one param has the same name: " + name);
+				}
+				seen.add(name);
+				this.originalParamTags.add(name);
 			}
-			String name = paramName.get();
-			if (seen.contains(name)) {
-				throw new DatastoreDataException(
-						"More than one param has the same name: " + name);
-			}
-			seen.add(name);
-			tags.add(name);
 		}
-		return tags;
+		return this.originalParamTags;
 	}
 
 	private GqlQuery<? extends BaseEntity> bindArgsToGqlQuery(String gql,
 			List<String> tags,
-			Object[] vals) {
+			List vals) {
 		Builder builder = GqlQuery.newGqlQueryBuilder(gql);
 		builder.setAllowLiteral(true);
-		if (tags.size() != vals.length) {
+		if (tags.size() != vals.size()) {
 			throw new DatastoreDataException("Annotated GQL Query Method "
 					+ this.queryMethod.getName() + " has " + tags.size()
-					+ " tags but a different number of parameter values: " + vals.length);
+					+ " initialTags but a different number of parameter values: " + vals.size());
 		}
 		for (int i = 0; i < tags.size(); i++) {
-			Object val = vals[i];
+			Object val = vals.get(i);
 			DatastoreNativeTypes.bindValueToGqlBuilder(builder, tags.get(i), val);
 		}
 		return builder.build();
 	}
+
+	private void resolveSpelTags(QueryTagValue queryTagValue) {
+		Expression[] expressions = getSpelExpressionParts();
+		StringBuilder sb = new StringBuilder();
+		Map<Object, String> valueToTag = new HashMap<>();
+		int tagNum = 0;
+		EvaluationContext evaluationContext = this.evaluationContextProvider
+				.getEvaluationContext(this.queryMethod.getParameters(),
+						queryTagValue.rawParams);
+		for (Expression expression : expressions) {
+			if (expression instanceof LiteralExpression) {
+				sb.append(expression.getValue(String.class));
+			}
+			else if (expression instanceof SpelExpression) {
+				Object value = expression.getValue(evaluationContext);
+				if (valueToTag.containsKey(value)) {
```

Comment:
use `computeIfAbsent()` instead of this `if`block

Reply:
Could you clarify? We want to do something different in both cases, not just only get-or-compute.